### PR TITLE
SCHED-97: Testing and fixing pre-Collector and Collector calculations

### PR DESCRIPTION
**Pull request by sraaphorst**
This PR comprises the calculations for the pre-Collector data based on the `time_grid` of the period we are interested in scheduling, and loads in the programs from a zip file, parses them using the OCS API, and populates the mini-model.

The `Collector`, once created, calculates the pre-Collector data, and then uses this data to calculate the information required for the program. This consists of populating the following fields in the `Collector`:

- `_programs` - A map (`dict`) of program IDs to the populated mini-model `Program` objects.
- `_observations` - A map (`dict`) of the mini-model enum `Site` keys to a list of the `Observation` objects for each site.
- `_target_info` - The primary information calculated here: a `dict` consisting of key `(TargetName, ObservationID, NightIndex)` to `TargetInfo`, which contains the necessary information about the position of the target for each `time_slot` during the indexed night, which is an index into the `time_grid` that comprises the JDay representation for each night. 

All pre-Collector and Collector data types and structures  can be found in the `collector/__init__.py` file:
- `NightEvents` - The events for a night (cached for a given `time_grid` and `Site`), calculated as per the Scheduler workflow document and represented using numpy / astropy types.
- `NightEventManager` - The caching mechanism that should be used to access `NightEvents` as required.
- `TargetInfo` - The information about a target, including its position (either as ephemerides or proper-motion adjusted coordinates) and visibility calculation information.

The way to invoke the `Collector` is via the following script:
```
scripts/run_collector.py
```

### CODE REVIEW

File: collector/__init__.py [hunk 1]
```diff
@@ -337,28 +345,34 @@ def _calculate_target_info(self,
         rem_visibility_frac_numerator = obs.exec_time() - obs.total_used()
 
         for ridx, jday in enumerate(reversed(self.time_grid)):
```
Comment (stroncod) [inline collector/__init__.py:347]: Are we calculating visibility for the whole timegrid it seems, does it takes too much time? I wonder as Bryan used to parallelize that to gain speed and maybe is something we might want to add in the future.

File: horizons/test/test_horizons_client.py [hunk 2]
```diff
@@ -1,5 +1,3 @@
-#import os, sys
-#sys.path.insert(0, os.path.abspath(".."))
 import pytest
```
[the remark this answers is not in the document]
Comment (sraaphorst) [reply horizons/test/test_horizons_client.py:1]: No worries. You have no idea how many times I've accidentally done things like this!

File: collector/__init__.py (hunk 1 above)
Comment (sraaphorst) [reply collector/__init__.py:347]: Yes, we plan to parallelize as we are calculating visibility across all base targets for all observations across the entire time grid, so it'll be essential when we work with a larger data set. I'll add a Jira task for that so we don't forget to add that.

File: environment.yml [hunk 3]
```diff
@@ -2,12 +2,12 @@ name: greedymax-env
 dependencies:
   - python=3.9.*
   - astropy
-  - numpy
+  - hypothesis
+  - joblib
+  - matplotlib
```
[the remark this answers is not in the document]
Comment (sraaphorst) [reply environment.yml:7]: 👍 I think, at least for now, Bryan would probably like to keep matplotlib in as he'll probably find some good uses for it at least during development, but sure, we should try to keep this file as clean as we can.

File: collector/__init__.py [hunk 4]
```diff
@@ -87,12 +87,11 @@ def __post_init__(self):
...
         # We want these as Python lists because the entries will have different lengths.
         # TODO: Thus, the np.vectorize approach will not work here, but we keep it here as documentation.
-        # self.utc_times = np.vectorize(lambda t: t.to_datetime('utc'))(self.times)
-        self.utc_times = [t.to_datetime('utc') for t in self.times]
+        self.utc_times = [t.to_datetime(pytz.UTC) for t in self.times]
         self.local_times = [t.to_datetime(self.site.value.timezone) for t in self.times]
         self.local_sidereal_times = [vskyutil.lpsidereal(t, self.site.value.location) for t in self.times]
 
-        def altazparang(pos: SkyCoord):
+        def altazparang(pos: List[SkyCoord]):  # -> Tuple[npt.NDArray[Angle], npt.NDArray[Angle], npt.NDArray[Angle]]:
```
[the remark this answers is not in the document]
Comment (sraaphorst) [reply collector/__init__.py:94]: Oh... it is actually that. This bit was a pain to write... the old code had me confused in the way that some of these things were being calculated so I had to muck around until I got the logic down properly. I'll put the typing back in and remove the unnecessary comments. Seems like we can't just return the `zip(...)` though or I get a whack of warnings.

**Pull request merged**